### PR TITLE
Refactor cleanupIDs

### PR DESCRIPTION
**Pull request by TrySound**
I still did get how id generation works but data flow is clear now at
least.

- covered with types
- migrated to visitor plugin api
- got rid of traverse api

### CODE REVIEW

File: plugins/cleanupIDs.js [hunk 1]
```diff
@@ -264,15 +108,190 @@ function generateID(currentID) {
...
+                id = match[1]; // href reference
+              }
+            }
+            if (id != null) {
+              let refs = referencesById.get(id);
+              if (refs == null) {
+                refs = [];
+                referencesById.set(id, refs);
+              }
+              refs.push({ element: node, name, value });
```
Comment (mrazauskas) [inline plugins/cleanupIDs.js:231]: Might be I missed something. Isn’t this the same, but slightly shorter:

```js
const refs = referencesById.get(id) || [];
if (!refs.length) {
  referencesById.set(id, refs);
}
refs.push({ element: node, name, value });
```

Comment (TrySound) [reply plugins/cleanupIDs.js:231]: Yes, though I prefer more explicit null check, length check is less obvious

Comment (mrazauskas) [reply plugins/cleanupIDs.js:231]: Sure. I was just looking around for examples. Working on a custom plugin. Just to say, Visitor API is a pleasure to use.

Comment (TrySound) [reply plugins/cleanupIDs.js:231]: Thanks. All plugins will use visitor api. full, perItem and perItemReverse will be removed in v3.

**Review by matheus1lva — APPROVED** [review]

**Pull request merged**